### DISCUSSION TITLE
Tiny Fpga Bx

### DISCUSSION THREAD

**Discussion by TurBoss** [multigcs/LinuxCNC-RIO#21] [Show and tell]
I just received an email that my pre-order (may) is shipped

Specifications
===========

Height: 1.4 inches, width: 0.7 inches
Programming interface: USB 2.0 full-speed (12 mbit/sec)
ICE40LP8K FPGA
    7,680 four-input look-up-tables
    128 KBit block RAM
    Phase Locked Loop
    41 user IO pins
8 MBit of SPI Flash
Onboard 3.3 V (300 mA) and 1.2 V (150 mA) LDO regulators
Low-Power 16 MHz MEMs Oscillator
    1.3 mA power when active
    50 ppm stability.

they sell it for 50€

https://www.crowdsupply.com/tinyfpga/tinyfpga-ax-bx

:)

**Comment by TurBoss** [comment]
Hello,

The board just arrived, looks awesome but I'm unable to program it with `make tinyprog`

got this error messages

```
[turboss@turbox Firmware]$ make tinyprog 
tinyprog -p rio.bin

    TinyProg CLI
    ------------
    Using device id 1d50:6130
    Only one board with active bootloader, using it.

    Programming /dev/ttyACM0 with rio.bin
Traceback (most recent call last):
  File "/home/turboss/venv_cnc/bin/tinyprog", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/turboss/Projects/TinyFPGA-Bootloader/programmer/tinyprog/__main__.py", line 404, in main
    addr = fpga.meta.userimage_addr_range()[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/turboss/Projects/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 271, in userimage_addr_range
    return self._get_addr_range(u"userimage")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/turboss/Projects/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 278, in _get_addr_range
    addr_map = self.root.get(u"bootmeta", {}).get(
               ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get'
make: *** [Makefile:30: tinyprog] Error 1
```


Any idea ?


Thanks

**Comment by multigcs** [comment]
Hmmm, don't know,
i flashed my tinyBx over SPI directly from the RPI, not over USB.

try
* 'tinyprog -l' to check if the bootloader works
* 'tinyprog --update-bootloader' to update the bootloader

**Comment by TurBoss** [comment]
hello,

I could rewrite the firmware using a tl866 but no a buspirate 4 (broken for flashrom)

since i can use the tl866 programer, its ok to upload the rio.bin? that will wipe the bootloader, i think is okay :)

thanks!

**Comment by multigcs** [comment]
initially I also flashed via USB, so it should work.

But yes, if you do it via SPI, the bootloader is gone.

I don't know the tl866, I went to the SPI ROM pins on the bottom of the board.

**Comment by TurBoss** [comment]
![imagen](https://github.com/multigcs/LinuxCNC-RIO/assets/184088/6dbc1275-10ae-4674-b247-3d7dcdb7f63b)


:) gona test it thank you

**Comment by TurBoss** [comment]
here are some notes I recopiled


BUS PIRATE
=========

(DOESNT WORK)

fthe v3 version of the hardware works with flashrom while the v4 doesn't

notes from : https://discourse.tinyfpga.com/t/bx-bootloader-bricked/852/9 (ssl cert is broken)

I just unbricked a TinyFPGA BX with a bus pirate and flashrom.
I soldered wires from the pads under the BX to the bus pirate connector like this :

```
BusPirate <--> BX

GND G of course
5V Vin (it's not a pad)
MISO IO1
CS SS
MOSI IO0
CLK SCK
```

I had to expand the firmware binfile to the flash size but adding zeros :

```
dd if=/dev/zero of=tinyfpga_bx_fw_1Mb.bin oflag=append conv=notrunc bs=1 count=749636
```

then flashrom was happy:

```
flashrom --programmer buspirate_spi:dev=/dev/ttyUSB0 -f -w ../tinyfpga_bx_fw_1Mb.bin
```

(flashrom compiled from source, packaged version on debian systems is usually too old)



TL866 II
=======
(WORKS)

notes from :https://discourse.tinyfpga.com/t/cant-program-tinyfpga-bx-anymore-solved/1384 (ssl cert is broken)


I used a TL866-II Plus programmer to re-flash the SPI Flash Memory on the board using the ISP port.

    Solder wires to the bottom side exposed pads of the tinyfpga SCK, IO1, IO0, SS; 3v, G (I used the one next to the 3v connector, not the ones at the bottom of the board)
    Plug in TL866-II Plus ISP adapter (the plug on the opposite end of the USB port)
    Connect the pins as noted (left side is TL866-II Plus, right side is tinyfpga)
    
    ```
    
        /CS -> SS
        VCC (red wire) -> 3v
        GND (black wire) -> G
        MOSI -> IO0
        MISO -> IO1 (optional, but recommended, also connect a 30pF capacitor between this and G)
        CLK -> SCK
    ```
    
    open Xgpro
    
    select IC AT25DF081 (the chip used is actually AT25SF081, but this is compatible)
    unselect Check ID (were going to manually check the id, since were using a different IC than the one were selecting)
    
    Press and hold Reset on TinyFPGA while you select Read ID from the Device menu (or press the read ID button on the toolbar)
    
    In the text box on the right side, it should say: Read Device ID: 1F 85 01 *** DO NOT CONTINUE IF THIS IS NOT DISPLAYED double check your wiring ***
    
    Select Device, Read, or press the Read button on the toolbar.
    
    Press and hold Reset on the Tinyfpga, click Read in Xgpro, hold the reset button for the duration of the read until it says Read finished!, this should take around 5 seconds.
    
    Click cancel, then save this as a backup.
    
    Download the tinyfpga bootloader binary from here: https://github.com/tinyfpga/TinyFPGA-Bootloader/releases/ 19 (BE SURE YOU GET tinyfpga_bx_fw.bin)
    
    From Xgpro, Open tinyfpga_tx_fw.bin from wherever you saved it
        set file format to binary
        region to Default
        load mode to normal
        From file start Addr to 00000
        To buffer start addr to 00000
        Clear buffer with 0xFF (this is the only non-default setting, but it shouldnt matter if you dont change it)
    
    Select program from device menu, or press the program button on the toolbar.
    
    While pressing and holding the TinyFPGA reset button (DO NOT LET GO UNTIL PROGRAMMING FINISHES), click the program button This may take 15-30 seconds to finish.
    
    unplug and try programming your device (it should be fixed)!

Important note press and hold the Reset button on the TinyFPGA for the entire duration that you are reading/verifying/etc!



MINIPRO
========



Using TL888 II from linux with minipro tool



substact `1048576` - `rio.bin size`
use the result as `count` in the dd command


```
cp rio.bin rio_1M.bin
dd if=/dev/zero of=tinyfpga_bx_fw_1Mb.bin oflag=append conv=notrunc bs=1 count=749636
minipro -i -p AT25SF081@TSSOP8 -w rio_1M.bin
```

Output should be like this :

```
minipro -i -p AT25SF081@TSSOP8 -w fw_1M.bin 
Found TL866II+ 04.2.132 (0x284)
Activating ICSP...
Chip ID: 0x1F8501  OK
Erasing... 2.01Sec OK
Writing Code...  8.25Sec  OK
Reading Code...  4.88Sec  OK
Verification OK
```


should be done

**Comment by TurBoss** [comment]
Hello, 

Got the usb to work 

had the "missing metatdta" issue

so i created a new virtualenv for the task, installed the PR 73 hack and ran tinyprog as python module

```bash
python3 -m venv venv
source venv/bin/activate
git clone https://github.com/drujensen/TinyFPGA-Bootloader.git -b dj/missing-metadata-crowd-supply-batch-hack
cd TinyFPGA-Bootloader/programer
pip install -e .
python -m tinyprog -l
```

should give something like this:

without press the button:

```
    TinyProg CLI
    ------------
    Using device id 1d50:6130
    No port was specified and no active bootloaders found.
    Activate bootloader by pressing the reset button.
```

and with the button pressed once

```

    TinyProg CLI
    ------------
    Using device id 1d50:6130
    Only one board with active bootloader, using it.

    Boards with active bootloaders:

        /dev/ttyACM0: TinyFPGA BX 0.0.0
            UUID: XXXXXXXX-XXXXXXX-XXXXXX
            FPGA: ICE40LP8K-cm81
```


:)

**Reply by degesz** [reply]
Hello, can you upload a bitstream to the TinyFPGA using this?
I tried this and while i could read the metadata as shown above, I couldn't program anything because it got stuck at 45% and said read error
`
(tinyprog-venv) root@degesz-laptop:/home/degesz/TinyFPGA-Bootloader/programmer# tinyprog -p /home/degesz/Documents/fpga/test-tiny.bin 

    TinyProg CLI
    ------------
    Using device id 1d50:6130
    Only one board with active bootloader, using it.

    Programming /dev/ttyACM0 with /home/degesz/Documents/fpga/test-tiny.bin
    Programming at addr 028000
    Waking up SPI flash
    135100 bytes to program
    Programming and Verifying:  45%|████████████████████████████████▋                                       | 61.4k/135k [00:01<00:01, 56.2kB/s]
Failed to read from serial port:
device reports readiness to read but returned no data (device disconnected or multiple access on port?)
`

Any ideas how to program? (the bitstream file was exported from IceStudio)

**Reply by multigcs** [reply]
this is not a problem with the bitsteam, but with the programmer, i using SPI to programm the flash.
the usb/serial programmer is buggy.
You can try another host system to programm, RPI -> X86, Linux -> Windows

**Comment by multigcs** [comment]
i wonder if it is possible to take the usb code and use it as an interface, at least for testing. USB and LinuxCNC are not the best friends :)
